### DISCUSSION TITLE
Where in the code is the ptx/sass loaded into GPU memory?

### DISCUSSION THREAD

**Discussion by ghost** [NVIDIA/open-gpu-kernel-modules#452] [Q&A]
Hello!
I am trying to figure out where in the driver code the PTX/SASS binary loaded/parsed into the GPU. Can someone point me to where this is done? I couldn't find any discussion on this.

Thank you!

**Comment by aaronp24** [comment]
GPU code that runs in graphics or compute contexts is loaded by the user-mode drivers, not the kernel driver code in this repository.

**Reply by ghost** [reply]
Thanks, does that mean also that the memory mapping and allocation for the code/data section of a kernel is also handled by the user mode drivers?

**Reply by aaronp24** [reply]
The user-mode drivers do call into the kernel module to create and memory allocations. See the `NV01_ALLOC_MEMORY`, `NV04_ALLOC`, `NV04_MAP_MEMORY`, and `NV04_MAP_MEMORY_DMA` operations. But those calls are pretty low-level. What are you looking to do, exactly?

**Reply by ghost** [reply]
I want to see if I can control which virtual address the code/data will be mapped to, whether that is different for each kernel or its uniform.
